File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var charSet = new Set();
    var l =0;
    var result =0;
    var res=0;
    
    for(var i=0; i<s.length; i++){
        while(charSet.has(s[i])){
            charSet.delete(s[l]);
            l=l+1;
        }
        charSet.add(s[i]);
        res = i-l+1;
        result = Math.max(result,res);
    }
    return result;
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l=0;
        int r=0;
        int max=0;
        
        HashSet<Character> hash_set = new HashSet<Character>();
        
        while(r<s.length()){
            if(!hash_set.contains(s.charAt(r))){
                hash_set.add(s.charAt(r));
                r++;
                max= Math.max(hash_set.size(), max);
            }else{
                hash_set.remove(s.charAt(l));
                l++;
            }
        }
        return max;
    }
}